Adding recipe to build a conda forge package for OOMMF

fangohr opened the pull request:
The Object Oriented MicroMagnetic Framework (OOMMF) forms a micromagnetics package (simulation of magnetic nano materials), and has its home page at http://math.nist.gov/oommf/

By making it available through conda forge, we hope to (i) help distribution and installation of the package and (ii) to provide a dependency for the Jupyter-OOMMF project, which provides a Jupyter notebook interface to OOMM (http://joommf.github.io).

For OSX and Linux, we build from a GitHub repository [1] that holds the source, whereas for Windows we take a pre-packaged zip file from the OOMMF creator. The code from the GitHub repository has a slight change, and thus we use a version number (ending in b1, for build 1) for Unix that differs from the Windows version number (ending in b0, for build 0).

[1] http://github.com/fangohr/oommf

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/oommf```) and found some lint.

Here's what I've got...


For **recipes/oommf**:

 * The license item is expected in the about section.
 * Selectors are suggested to take a ``<two spaces>#<one space>[<expression>]`` form.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/oommf```) and found it was in an excellent condition.

isuruf commented on recipes/oommf/build.sh:
```diff
@@ -0,0 +1,30 @@
+export OOMMF_TCL_CONFIG=${PREFIX}/lib/tclConfig.sh
+export OOMMF_TK_CONFIG=${PREFIX}/lib/tkConfig.sh
+cd oommf
+tclsh${TCLTKVERSION} oommf.tcl pimake distclean
```
You can use `tclsh` without the version number

fangohr replied:
Thanks. There is currently some confusion over tclsh/tk8.5 and tclsh/tk8.6 being build differently (with threading enabled or without), so I am not sure this is actually the same. Is there any update on this, or recommendation on how to proceed here, @takluyver ?

isuruf replied:
That may matter in the meta.yaml, but in the build script you need not worry about that

takluyver replied:
I think that conda can only have one version of tclsh installed, and because we specify the dependency, it should always be the correct one.

fangohr replied:
Thanks both - I have removed all references to TCLTKVERSION in the build.sh file; seems we can do without. That's a nice simplification.

fangohr replied:
It seems we run into a problem here, and I think putting the TCLTKVERSION  back into this bash script. Here are the observations:

## the error
<pre>
(testoommf) escape:~ fangohr$ oommf
<25728> oommf.tcl 1.2.1.0 Oc_Config warning:
Tcl version mismatch:
	/Users/fangohr/anaconda3/envs/testoommf/lib/tclConfig.sh from 8.5.18
	Running Tcl 8.5.9
(testoommf) escape:~ fangohr$ <25732> /Users/fangohr/anaconda3/envs/testoommf/opt/oommf/app/omfsh/darwin/omfsh 1.2.1.0 panic:
Tcl initialization error:

Can't find a usable init.tcl in the following directories: 
    /System/Library/Frameworks/Tcl.framework/Versions/8.5/Resources/Scripts /opt/anaconda1anaconda2anaconda3/lib/tcl8.5 /Users/fangohr/anaconda3/envs/testoommf/opt/oommf/app/omfsh/lib/tcl8.5 /Users/fangohr/anaconda3/envs/testoommf/opt/oommf/app/lib/tcl8.5 /Users/fangohr/anaconda3/envs/testoommf/opt/oommf/app/omfsh/library /Users/fangohr/anaconda3/envs/testoommf/opt/oommf/app/library /Users/fangohr/anaconda3/envs/testoommf/opt/oommf/app/tcl8.5.18/library /Users/fangohr/anaconda3/envs/testoommf/opt/oommf/tcl8.5.18/library

/System/Library/Frameworks/Tcl.framework/Versions/8.5/Resources/Scripts/init.tcl: version conflict for package "Tcl": have 8.5.18, need exactly 8.5.9
version conflict for package "Tcl": have 8.5.18, need exactly 8.5.9
    while executing
"package require -exact Tcl 8.5.9"
    (file "/System/Library/Frameworks/Tcl.framework/Versions/8.5/Resources/Scripts/init.tcl" line 20)
    invoked from within
"source /System/Library/Frameworks/Tcl.framework/Versions/8.5/Resources/Scripts/init.tcl"
    ("uplevel" body line 1)
    invoked from within
"uplevel #0 [list source $tclfile]"


This probably means that Tcl wasn't installed properly.


As a workaround, set the environment variable
TCL_LIBRARY to the name of a directory
containing an error-free init.tcl.
</pre>

## Can reproduce with the last line of start up script:
<pre>
(testoommf) escape:~ fangohr$ which oommf
/Users/fangohr/anaconda3/envs/testoommf/bin/oommf
(testoommf) escape:~ fangohr$ cat `which oommf`
#! /bin/bash
export OOMMF_TCL_CONFIG=/Users/fangohr/anaconda3/envs/testoommf/lib/tclConfig.sh
export OOMMF_TK_CONFIG=/Users/fangohr/anaconda3/envs/testoommf/lib/tkConfig.sh
tclsh /Users/fangohr/anaconda3/envs/testoommf/opt/oommf/oommf.tcl "$@"
</pre>

and then

<pre>
(testoommf) escape:~ fangohr$ tclsh /Users/fangohr/anaconda3/envs/testoommf/opt/oommf/oommf.tcl "$@"
(testoommf) escape:~ fangohr$ <25889> /Users/fangohr/anaconda3/envs/testoommf/opt/oommf/app/omfsh/darwin/omfsh 1.2.1.0 panic:
Tcl initialization error:

Can't find a usable init.tcl in the following directories: 
    /System/Library/Frameworks/Tcl.framework/Versions/8.5/Resources/Scripts /opt/anaconda1anaconda2anaconda3/lib/tcl8.5 /Users/fangohr/anaconda3/envs/testoommf/opt/oommf/app/omfsh/lib/tcl8.5 /Users/fangohr/anaconda3/envs/testoommf/opt/oommf/app/lib/tcl8.5 /Users/fangohr/anaconda3/envs/testoommf/opt/oommf/app/omfsh/library /Users/fangohr/anaconda3/envs/testoommf/opt/oommf/app/library /Users/fangohr/anaconda3/envs/testoommf/opt/oommf/app/tcl8.5.18/library /Users/fangohr/anaconda3/envs/testoommf/opt/oommf/tcl8.5.18/library

/System/Library/Frameworks/Tcl.framework/Versions/8.5/Resources/Scripts/init.tcl: version conflict for package "Tcl": have 8.5.18, need exactly 8.5.9
version conflict for package "Tcl": have 8.5.18, need exactly 8.5.9
    while executing
"package require -exact Tcl 8.5.9"
    (file "/System/Library/Frameworks/Tcl.framework/Versions/8.5/Resources/Scripts/init.tcl" line 20)
    invoked from within
"source /System/Library/Frameworks/Tcl.framework/Versions/8.5/Resources/Scripts/init.tcl"
    ("uplevel" body line 1)
    invoked from within
"uplevel #0 [list source $tclfile]"


This probably means that Tcl wasn't installed properly.


As a workaround, set the environment variable
TCL_LIBRARY to the name of a directory
containing an error-free init.tcl.
</pre>

## Fix
But using `tclsh8.5` instead of `tclsh` fixes the issue
<pre>
(testoommf) escape:~ fangohr$ tclsh8.5 /Users/fangohr/anaconda3/envs/testoommf/opt/oommf/oommf.tcl "$@"
(testoommf) escape:~ fangohr$ 2017-05-29 19:16:12.234 omfsh[25895:2080119]
</pre>
(This out output is good news; oommf working correctly.)

## diagnostics
<pre>
(testoommf) escape:~ fangohr$ which tclsh8.5
/Users/fangohr/anaconda3/envs/testoommf/bin/tclsh8.5
(testoommf) escape:~ fangohr$ which tclsh
/usr/bin/tclsh
(testoommf) escape:~ fangohr$ 
</pre>
This seems to suggest that tcl8.5 does not provide a symbolic link to `tclsh`, and that on this installation, the system tclsh is used instead of the conda one (resulting in the error).

## more info
<pre>
(testoommf) escape:~ fangohr$ conda info
Current conda install:

               platform : osx-64
          conda version : 4.2.13
       conda is private : False
      conda-env version : 4.2.13
    conda-build version : 2.1.13
         python version : 3.6.0.final.0
       requests version : 2.12.4
       root environment : /Users/fangohr/anaconda3  (writable)
    default environment : /Users/fangohr/anaconda3/envs/testoommf
       envs directories : /Users/fangohr/anaconda3/envs
          package cache : /Users/fangohr/anaconda3/pkgs
           channel URLs : https://repo.continuum.io/pkgs/free/osx-64
                          https://repo.continuum.io/pkgs/free/noarch
                          https://repo.continuum.io/pkgs/pro/osx-64
                          https://repo.continuum.io/pkgs/pro/noarch
            config file : /Users/fangohr/.condarc
           offline mode : False

(testoommf) escape:~ fangohr$ conda list 
# packages in environment at /Users/fangohr/anaconda3/envs/testoommf:
#
oommf                     1.2b0_20160930b1               0    conda-forge
openssl                   1.0.2l                        0  
pip                       9.0.1                    py36_1  
python                    3.6.1                         2  
readline                  6.2                           2  
setuptools                27.2.0                   py36_0  
sqlite                    3.13.0                        0  
tk                        8.5.18                        0  
wheel                     0.29.0                   py36_0  
xz                        5.2.2                         1  
zlib                      1.2.8                         3  
(testoommf) escape:~ fangohr$ 
</pre>

## Output from creating that environment `testoommf`
<pre>
escape:~ fangohr$ conda create -n testoommf python=3
Fetching package metadata .......
Solving package specifications: ..........

Package plan for installation in environment /Users/fangohr/anaconda3/envs/testoommf:

The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    openssl-1.0.2l             |                0         3.0 MB

The following NEW packages will be INSTALLED:

    openssl:    1.0.2l-0     
    pip:        9.0.1-py36_1 
    python:     3.6.1-2      
    readline:   6.2-2        
    setuptools: 27.2.0-py36_0
    sqlite:     3.13.0-0     
    tk:         8.5.18-0     
    wheel:      0.29.0-py36_0
    xz:         5.2.2-1      
    zlib:       1.2.8-3      

Proceed ([y]/n)? y

Fetching packages ...
openssl-1.0.2l 100% |####################################################################################################| Time: 0:00:01   2.70 MB/s
Extracting packages ...
[      COMPLETE      ]|#######################################################################################################################| 100%
Linking packages ...
[      COMPLETE      ]|#######################################################################################################################| 100%
#
# To activate this environment, use:
# > source activate testoommf
#
# To deactivate this environment, use:
# > source deactivate testoommf
#

escape:~ fangohr$ source activate testoommf
(testoommf) escape:~ fangohr$ conda install -c conda-forge oommf
Fetching package metadata .........
Solving package specifications: ..........

Package plan for installation in environment /Users/fangohr/anaconda3/envs/testoommf:

The following NEW packages will be INSTALLED:

    oommf: 1.2b0_20160930b1-0 conda-forge

Proceed ([y]/n)? y

Linking packages ...
[      COMPLETE      ]|#######################################################################################################################| 100%
(testoommf) escape:~ fangohr$ which oommf
/Users/fangohr/anaconda3/envs/testoommf/bin/oommf
</pre>

## Now what?
So, should we go back to 
`+tclsh${TCLTKVERSION} oommf.tcl pimake distclean`
instead of 
`+tclsh oommf.tcl pimake distclean`?

Or 'fix'(?) the conda `tk` package? [This seems risky to me.]

Or have I missed another origin of this issue?

Views, @isuruf, @takluyver ?

takluyver replied:
Yep, my conda installation also appears to have `tclsh8.5` and not `tclsh`, so I think we should either go back to using `tclsh8.5` here, or adjust the conda packages to also install it as `tclsh`.

isuruf replied:
What does `hash -r && which tclsh` give you?

fangohr replied:
<pre>
(testoommf) bash-3.2$ hash -r && which tclsh
/usr/bin/tclsh
(testoommf) bash-3.2$ which tclsh
/usr/bin/tclsh
(testoommf) bash-3.2$ 
</pre>

isuruf replied:
Ah, you are using the tcl package from defaults which doesn't have the symlink

fangohr replied:
You mean that we use the default package from conda? Does conda-forge have a better one? If so, how do we choose to use that? (Sorry for naive questions, very new to conda-forge.)

[The tclsh in the path in the example above is provided by OSX, I think.]

isuruf replied:
Package from the defaults channel has `tclsh8.5` only
Conda-forge package has `tclsh8.5` and `tclsh`

You can give higher priority to conda-forge to use conda-forge packages. This has to be done by the user though.

fangohr replied:
Ah, I see - thank you. 

In that case, I prefer to stick to hardcoding tclsh8.5 in our execution script: our users are likely to be beginners with conda, and will not be able to give higher priority to conda-forge over conda: if they can install something that looks like anaconda, and then install oommf using a line we give them (such as `conda install -c conda-forge oommf`), then that is already an achievement. If it then fails in a subtle way (because the have the conda tk and not conda-forge tk), that would be counterproductive.

I hope that makes sense and is agreeable?

isuruf replied:
```
conda config --add channels conda-forge
conda install oommf
```
Above should work. I think since there are incompatibilities between defaults and conda-forge, you should recommend the above command. `conda install -c conda-forge <pkg>` might not work sometimes. Having said that, using `tclsh8.5` makes sense if you want to keep compatibility with defaults.

fangohr replied:
Okay, thank you. 

I have updated the oommf-feedstack to include the version number, and will also follow your advice on recommending to users to add the condo-forge channel when they install.

isuruf commented on recipes/oommf/meta.yaml:
```diff
@@ -0,0 +1,58 @@
+# on Windows, we install from precompiled binary packages
+source:  # [unix]
+  git_url: https://github.com/fangohr/oommf  # [unix]
+  git_rev: 1.2b0_20160930b1  # [unix]
```
See http://conda-forge.github.io/docs/meta.html#build-from-tarballs-not-repos

Use
```yaml
source:
   url: https://github.com/fangohr/oommf/archive/1.2b0_20160930b1.tar.gz  # [unix]
   url: http://math.nist.gov/oommf/dist/oommf12b0_20160930_86_x64.zip     # [win]
   sha256: <unixsha256>  # [unix]
   sha256: <windowssha256>  # [win]
```

fangohr replied:
Thank you - good point. I have updated the recipe and modified the pull request.

isuruf commented on recipes/oommf/meta.yaml:
```diff
@@ -0,0 +1,58 @@
+
+test:  # [unix]
+  commands:  # [unix]
+    - oommf +version  # [unix]
```
This doesn't work in windows?

fangohr replied:
No. @takluyver has the details.

takluyver replied:
The `oommf` command line interface requires `tclsh`, which doesn't seem to be there on Windows in the conda packages of tcl/tk. For our Python wrapper code which calls OOMMF, we've found a way of [calling what we need](https://github.com/joommf/oommfc/blob/943ddc0476fde2f8cc34629687636a569fccbc4f/oommfc/oommf.py#L117) without the `oommf` command.

isuruf replied:
Have you tried using `m2w64-tcl` package from defaults?

takluyver replied:
We may well have missed that. I don't know if the precompiled oommf binaries for Windows are likely to be compatible with mingw/msys stuff. I've never really understood the compatibility story with the various Unix-tools-on-Windows systems.

isuruf replied:
True. Are the precompiled oommf binaries for Windows built with MSVC?

takluyver replied:
I don't know, unfortunately. @fangohr or @mb1a15 might have a better idea, but I don't think any of us are familiar with Windows.

rpep replied:
@isuruf I think they are, but not 100%.

isuruf commented on recipes/oommf/meta.yaml:
```diff
@@ -0,0 +1,58 @@
+    This version contains a bulk dmi extension  # [unix]
+    (https://github.com/fangohr/oommf/blob/master/oommf/app/oxs/local/BulkDMI.h)  # [unix]
+    which is not included in the OOMMF binaries and tarballs from NIST  # [unix]
+    (yet).  # [unix]
```
I don't think the selectors work here.

fangohr replied:
I heard rumours they might (i.e. that line selection takes place before the yaml is interpreted). Can we try if this works, and remove it if doesn't?

isuruf commented on recipes/oommf/meta.yaml:
```diff
@@ -0,0 +1,58 @@
+    - mb1a15
+    - takluyver
+    - davidcortesortuno
+    - rpep
```
@fangohr, have you asked these people permission to add them as maintainers?

fangohr replied:
Yes, thanks for checking.

isuruf commented on recipes/oommf/meta.yaml:
```diff
@@ -0,0 +1,58 @@
+
+requirements:
+  build:
+    - tk 8.5*  # [unix]
```
Add `toolchain` as a build dependency whenever the package has any compilation

fangohr replied:
Okay, done.

isuruf commented on recipes/oommf/build.sh:
```diff
@@ -0,0 +1,30 @@
+#!/bin/bash
+
+TCLTKVERSION=8.5
+CONDA_ROOT_PATH=$(conda info --root)
```
You can remove this

fangohr replied:
@takluyver - are we using $CONDA_ROOT_PATH for JOOMMF to find the location of conda, to then find OOMMF, or do you think this can be removed?

isuruf replied:
Even if it was, declaring it here is useless

takluyver replied:
Yep, this can safely be removed.

fangohr replied:
Done.

isuruf commented on recipes/oommf/build.sh:
```diff
@@ -0,0 +1,30 @@
+tclsh${TCLTKVERSION} oommf.tcl pimake distclean
+tclsh${TCLTKVERSION} oommf.tcl pimake upgrade
+tclsh${TCLTKVERSION} oommf.tcl pimake
+tclsh${TCLTKVERSION} oommf.tcl +platform
```
Are there any tests we can run after the compilation?

fangohr replied:
The `tclsh${TCLTKVERSION} oommf.tcl +platform` is a little bit like a test: it spits out diagnostic information about this install. (Another command we could run, and which is listed in the `test` section in `meta.yaml', is `tclsh${TCLTKVERSION} oommf.tcl +version`, which reports just the version number, but is also a test to check some basics of the installation.

There is no automatic test suit for OOMMF.

isuruf replied:
> There is no automatic test suit for OOMMF.

Ah, then okay.

isuruf commented on recipes/oommf/bld.bat:
```diff
@@ -0,0 +1,5 @@
+curl -o oommf.zip  "http://math.nist.gov/oommf/dist/oommf12b0_20160930_86_x64.zip"
```
This is downloaded and extracted here because of https://github.com/conda-forge/staged-recipes/pull/3038/files#diff-e41816a448c34eb6ab6f515c151653a2R10. You don't need to download it again here.

fangohr replied:
So does that mean I can remove all the lines from the `bld.bat` file:
<pre>
curl -o oommf.zip  "http://math.nist.gov/oommf/dist/oommf12b0_20160930_86_x64.zip"

mkdir %PREFIX%\opt

7za x -o%PREFIX%\opt oommf.zip
</pre>
? I presume so and will try.

takluyver replied:
I think you'll still need to copy the files into `%PREFIX%\opt\oommf`

fangohr replied:
But @isuruf said that the files are 'downloaded and extracted'. So we don't need the unzip command, but instead need to move all unzipped files into `/opt`?

isuruf replied:
Yes, you just need to move the files to `%PREFIX%\opt`

fangohr replied:
Okay, so when oommf12b0_20160930_86_x64.zip is unzipped, it creates a subdirectory `oommf` which contains all files. My understanding is that the whole directory needs to go to `%PREFIX%\opt`.

So the right unix command would be 
- `mv oommf %PREFIX%\opt` or
- `mkdir %PREFIX%\opt/oommf && mv oommf/* %PREFIX%\opt/oommf`

As I haven't got a windows machine for testing, can somebody suggest a suitable command to achieve this in reply to this comment? (@takluyver, @rpep, @mb1a15).

isuruf replied:
```
mkdir %PREFIX%\opt
move oomf %PREFIX%\opt
```

takluyver replied:
Looks like the [move command](https://ss64.com/nt/move.html) is equivalent to `mv` in Unix.

fangohr replied:
Done, thanks.

fangohr commented on recipes/oommf/meta.yaml:
```diff
@@ -0,0 +1,46 @@
+    - mb1a15
+    - takluyver
+    - davidcortesortuno
+    - rpep
```
I think we have addressed all queries now. Thank you @isuruf for the quick review of this, the thorough review, and the repeated feedback and help.

isuruf commented:
Conda build will check if the extracted tarball has a single folder and if so moves the contents. So something like following should work.
```
mkdir %PREFIX%\opt\oommf
move * %PREFIX%\opt\oommf
```

fangohr commented:
Thanks - I replaced the `move` with an `xcopy`: I think the `move` moved out `bld.bat` script as well, and then upset conda. The `xcopy` copies subdirectories as well `/e`, which `copy` doesn't do, it appears.

I think we have now addressed all issues, and the builds seem to be working. Many thanks, @isuruf .

isuruf commented:
Thanks for the PR.
@davidcortesortuno, @mb1a15, welcome to conda-forge. See https://github.com/conda-forge/oommf-feedstock#updating-oommf-feedstock for the conda-forge workflow.
